### PR TITLE
webdriverIO appium

### PR DESCRIPTION
**Pull request by shubhamyadav-bs**
- Added browserstack single run
- Added browserstack parallel run
- Added browserstack parallel device run

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+expectChai(x).to.eql(y);
+console.log('hello');
+}
+function mysort() {
+    let xx = [399, 429, 499,  499,  549, 599, 599, 599,  649,  699, 699, 799, 899,  899,  899, 899, 999, 999, 1199, 1299, 1399 ]
+    console.dir(xx.slice().sort(function(a, b){return a-b}));
+}
+//validate();
+
+//mysort();
```
Comment (samirans89) [inline index.js]: needed?

File: resources/conf/wdio-bstack-local-parallel-devices.conf.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+var overrides = {
+  onBrowserstack: true,
+
+  user: process.env.BROWSERSTACK_USER,
+  key: process.env.BROWSERSTACK_ACCESSKEY,
+
+  specs: ["./test/specs/local/local.spec.js"],
+
+  capabilities: [
```
Comment (samirans89) [inline resources/conf/wdio-bstack-local-parallel-devices.conf.js:7]: can we use commonCapabilities for values of build, project etc and all other caps which are truly common for the test run?

an example we had created for a customer:

https://github.com/gregjopa/webdriver-io-boilerplate/blob/main/resources/config/wdio.prod.browserstack.conf.ts#L16

File: resources/conf/wdio-bstack-local-parallel-devices.conf.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        []
+      );
+    } else {
+      await driver.takeScreenshot();
+      await browser.executeScript(
+        'browserstack_executor: {"action": "setSessionStatus", "arguments": {"status":"failed","reason": "At least 1 assertion failed"}}',
+        []
+      );
+    }
+  },
```
Comment (samirans89) [inline resources/conf/wdio-bstack-local-parallel-devices.conf.js]: this segment (test hook) is getting repeated in a number of files.

we could create this file extends wdio-bstack.conf.js extends wdio.conf.js.

and then have bstack common hooks in wdio-bstack.conf.js. 3 level hierarchy.

File: resources/conf/wdio-bstack-local-parallel-devices.conf.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    if (passed) {
+      await driver.executeScript(
+        'browserstack_executor: {"action": "setSessionStatus", "arguments": {"status":"passed","reason": "Assertions passed"}}',
+        []
+      );
+    } else {
+      await driver.takeScreenshot();
+      await browser.executeScript(
+        'browserstack_executor: {"action": "setSessionStatus", "arguments": {"status":"failed","reason": "At least 1 assertion failed"}}',
```
Comment (samirans89) [inline resources/conf/wdio-bstack-local-parallel-devices.conf.js]: many customers struggle to get the actual error reason on the dashboard.

Check this.
https://github.com/gregjopa/webdriver-io-boilerplate/blob/main/resources/config/wdio.prod.browserstack.conf.ts#L90-L98


```
      const reason = (
        (error).toString().replace(/[^a-zA-Z0-9.]/g, " ")
      ).substring(0, 255);
      await browser.executeScript(
        'browserstack_executor: {"action": "setSessionStatus", "arguments": {"status":"failed","reason": "' +
          reason +
          '"}}'
      );
```

File: resources/conf/wdio-bstack-single.conf.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      app: process.env.BROWSERSTACK_ANDROID_APP_ID,
+      autoGrantPermissions: true,
+      platformName: "Android",
+      project: "browserstack-examples-appium-webdriverio",
+    },
+  ],
+
+  logLevel: "warn",
+  bail: 0,
+  baseUrl: "http://localhost",
```
Comment (samirans89) [inline resources/conf/wdio-bstack-single.conf.js]: is localhost relevant for single conf?

File: resources/conf/wdio-onprem-suite.conf.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    //   automationName: "XCUITest",
+    //   platformName: "iOS",
+    // },
+  ],
+
+  logLevel: "warn",
+  // logLevels: {
+  //     webdriver: 'info',
+  //     '@wdio/appium-service': 'info'
+  // },
```
Comment (samirans89) [inline resources/conf/wdio-onprem-suite.conf.js]: lets remove un-required / commented logic

File: test/pages/page.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    //     options: {
+    //       x: endX,
+    //       y: endY,
+    //     },
+    //   },
+    //   {
+    //     action: "release",
+    //     options: {},
+    //   },
+    // ]);
```
Comment (samirans89) [inline test/pages/page.js]: remove commented logic

**Review by samirans89 — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (shubhamyadav-bs) [reply index.js]: removed

File: resources/conf/wdio-bstack-local-parallel-devices.conf.js (hunk 2 above)
Comment (shubhamyadav-bs) [reply resources/conf/wdio-bstack-local-parallel-devices.conf.js:7]: implemented similar commonCapabilities

File: resources/conf/wdio-bstack-local-parallel-devices.conf.js (hunk 3 above)
Comment (shubhamyadav-bs) [reply resources/conf/wdio-bstack-local-parallel-devices.conf.js]: `wdio.conf.js` will be extended by `wdio-bstack.conf.js` which will in turn be extended by individual confs.

File: resources/conf/wdio-bstack-single.conf.js (hunk 5 above)
Comment (shubhamyadav-bs) [reply resources/conf/wdio-bstack-single.conf.js]: removed

File: resources/conf/wdio-onprem-suite.conf.js (hunk 6 above)
Comment (shubhamyadav-bs) [reply resources/conf/wdio-onprem-suite.conf.js]: removed

File: test/pages/page.js (hunk 7 above)
Comment (shubhamyadav-bs) [reply test/pages/page.js]: removed

File: resources/conf/wdio-bstack-local-parallel-devices.conf.js (hunk 4 above)
Comment (shubhamyadav-bs) [reply resources/conf/wdio-bstack-local-parallel-devices.conf.js]: added reason

**Review by NightWing1998 — APPROVED** [review]

**Review by samiran-bs — APPROVED** [review]

**Review by samiran-bs — APPROVED** [review]

**Review by abhi-singhs — APPROVED** [review]

**Pull request merged**